File: lint-arwaky/archive/agent/arch_compliance_orchestrator.py
"""architecture_orchestrator — Logic for resolving ArchitectureConfig into LayerDefinitions."""

from __future__ import annotations

from ..taxonomy import (
    ArchitectureConfig,
    LayerDefinition,
    ArchitectureRule,
    LAYER_GLOBAL,
    LayerMapVO,
)
from ..contract import ArchitectureOrchestratorAggregate


class ArchitectureOrchestrator(ArchitectureOrchestratorAggregate):
    """Orchestrates the resolution of architectural configurations."""

    def resolve_effective_layer_map(self, config: ArchitectureConfig) -> LayerMapVO:
        """
        Merges base layer definitions with global and layer-specific rules.
        Replaces the logic previously found in ArchitectureConfig.get_effective_layer_map().
        """
        # Ground mandatory imports
        _arch_aggregate: ArchitectureOrchestratorAggregate = self

        layer_map = self._resolve_layers(config)
        layer_map = self._merge_layer_rules(layer_map, config)
        return LayerMapVO(values=self._finalize_map(layer_map))

    def _resolve_layers(self, config: ArchitectureConfig):
        """Initialize base layer definitions from config — preserve suffix policy & allowed/forbidden suffixes AS-IS."""
        layer_map = {}
        for name, entry in config.layers.items():
            layer_map[str(name)] = entry
        return layer_map

    def _merge_layer_rules(self, layer_map, config: ArchitectureConfig):
        """Merge all rules from config into the layer map, handling global, specialized, and base scopes."""
        result = dict(layer_map)  # shallow copy to avoid mutating caller's dict

        for rule in config.rules:
            scope = str(rule.scope)
            if scope == LAYER_GLOBAL:
                self._apply_global_rule(result, rule)
                continue

            base_name = scope.split("(")[0] if "(" in scope else scope
            base_definition = result.get(base_name)
            if base_definition is None:
                continue  # omit rules targeting unknown layers

            # Clone specialized scope from base definition if not yet present
            if scope not in result and scope != base_name:
                result[scope] = base_definition.model_copy(deep=True)

            if "(" in scope:
                self._apply_specialized_rule(result, scope, rule)
            else:
                self._apply_base_rule(result, scope, rule)

        return result

    def _apply_global_rule(self, result, rule: ArchitectureRule) -> None:
        """Apply rule with scope=LAYER_GLOBAL to all non-specialized layers."""
        for layer_name in list(result.keys()):
            if "(" in layer_name:
                continue  # omit specialized layers for global rules
            result[layer_name] = self._merge_rule_into_definition(result[layer_name], rule)

    def _apply_specialized_rule(self, result, scope, rule: ArchitectureRule) -> None:
        """Apply rule to a single specialized layer scope (e.g. 'service(admin)')."""
        if scope in result:
            result[scope] = self._merge_rule_into_definition(result[scope], rule)

    def _apply_base_rule(self, result, scope, rule: ArchitectureRule) -> None:
        """Apply rule to a single base/named layer (no parentheses in scope)."""
        if scope in result:
            result[scope] = self._merge_rule_into_definition(result[scope], rule)

    def _finalize_map(self, layer_map):
        """Final pass over resolved layer map — ready for return."""
        return layer_map

    def _get_target_layers(self, rule: ArchitectureRule, layer_map):
        scope = str(rule.scope)
        if scope == LAYER_GLOBAL:
            return list(layer_map.keys())

        # Handle list of layers in scope
        if hasattr(rule.scope, 'values'): # PatternList
            return [str(s) for s in rule.scope.values if str(s) in layer_map]

        if scope in layer_map:
            return [scope]

        return []

    def _merge_rule_into_definition(self, definition: LayerDefinition, rule: ArchitectureRule) -> LayerDefinition:
        """Surgically merges rule settings into a LayerDefinition."""
        update_data = {}
        update_data.update(self._apply_rule_overrides(rule))
        update_data.update(self._apply_exceptions(rule))

        # Additively merge exceptions instead of overwriting!
        from ..taxonomy import PatternList
        existing_exceptions = definition.exceptions.values if definition.exceptions else []
        rule_exceptions = rule.exceptions.values if rule.exceptions else []
        if rule_exceptions:
            merged_exceptions = list(sorted(set(existing_exceptions + rule_exceptions)))
            update_data["exceptions"] = PatternList(values=merged_exceptions)

        return definition.model_copy(update=update_data)

    def _apply_rule_overrides(self, rule: ArchitectureRule):
        """Extract scalar override fields from a rule (simple fields, booleans, violation messages)."""
        update_data = {}

        # 1. Simple scalar fields (override if not None)
        simple_fields = [
            "word_count", "suffix_policy",
            "mandatory_import_violation_message", "forbidden_import_violation_message",
            "word_count_violation_message", "suffix_violation_message",
            "no_primitives_violation_message", "min_lines_violation_message",
            "max_lines_violation_message", "barrel_completeness_violation_message",
            "forbid_internal_all_violation_message", "forbidden_bypass_violation_message",
            "mandatory_class_definition_violation_message", "dead_inheritance_bypass_violation_message",
            "orphan_violation_message", "check_unused_mandatory_imports_violation_message",
            "min_lines", "max_lines"
        ]

        # Role-specific violation messages
        role_violation_fields = [
            "no_domain_logic_violation_message", "must_implement_service_container_aggregate_violation_message",
            "lazy_eager_initialization_only_violation_message", "stateless_execution_violation_message",
            "single_execution_goal_violation_message", "high_level_policy_only_violation_message",
            "coordinates_multiple_orchestrators_violation_message", "crud_only_violation_message",
            "no_decision_logic_violation_message", "thread_async_safe_violation_message",
            "no_domain_data_storage_violation_message", "owns_system_health_transitions_violation_message",
            "lifecycle_tracking_only_violation_message",
            "forbid_any_type_violation_message"
        ]

        for field in simple_fields + role_violation_fields:
            val = getattr(rule, field, None)
            if val is not None:
                update_data[field] = val

        # 2. Boolean/Status fields (override if not None)
        boolean_fields = [
            "no_primitives", "barrel_completeness", "forbid_internal_all",
            "mandatory_class_definition", "dead_inheritance_bypass", "check_orphan",
            "check_unused_mandatory_imports", "no_domain_logic",
            "must_implement_service_container_aggregate", "lazy_eager_initialization_only",
            "stateless_execution", "single_execution_goal", "high_level_policy_only",
            "coordinates_multiple_orchestrators", "crud_only", "no_decision_logic",
            "thread_async_safe", "no_domain_data_storage", "owns_system_health_transitions",
            "lifecycle_tracking_only",
            "forbid_any_type"
        ]

        for field in boolean_fields:
            val = getattr(rule, field, None)
            if val is not None:
                update_data[field] = val

        return update_data

    def _apply_exceptions(self, rule: ArchitectureRule):
        """Extract collection/exception fields from a rule (imports, suffixes, bypasses, etc.)."""
        update_data = {}
        collection_fields = [
            "allowed_import", "forbidden_import", "mandatory_import",
            "allowed_suffix", "forbidden_suffix", "exceptions",
            "forbidden_bypass", "orphan_entry_points",
            "mandatory_imports", "forbidden_bypass_custom_messages",
            "dead_inheritance_bypass_custom_messages"
        ]

        for field in collection_fields:
            val = getattr(rule, field, None)
            if val is not None:
                if hasattr(val, 'values') and val.values or isinstance(val, list) and val:
                    update_data[field] = val

        return update_data

File: lint-arwaky/src-rust/agent/arch_compliance_orchestrator.rs
// arch_compliance_orchestrator — Logic for resolving ArchitectureConfig into LayerDefinitions.
use std::collections::HashMap;
use crate::contract::ArchitectureOrchestratorAggregate;
use crate::taxonomy::{
    ArchitectureConfig, ArchitectureRule, LayerDefinition, LayerMapVO, LayerNameVO,
    LAYER_GLOBAL, BooleanVO, ErrorMessage,
};

pub struct ArchitectureOrchestrator;

impl ArchitectureOrchestratorAggregate for ArchitectureOrchestrator {}

impl ArchitectureOrchestrator {
    pub fn new() -> Self {
        Self
    }

    /// Merges base layer definitions with global and layer-specific rules.
    pub fn resolve_effective_layer_map(&self, config: &ArchitectureConfig) -> LayerMapVO {
        let mut layer_map = self.resolve_layers(config);
        layer_map = self.merge_layer_rules(layer_map, config);
        LayerMapVO { values: layer_map }
    }

    fn resolve_layers(&self, config: &ArchitectureConfig) -> HashMap<LayerNameVO, LayerDefinition> {
        config.layers.clone()
    }

    fn merge_layer_rules(
        &self,
        mut result: HashMap<LayerNameVO, LayerDefinition>,
        config: &ArchitectureConfig,
    ) -> HashMap<LayerNameVO, LayerDefinition> {
        for rule in &config.rules {
            let scope = rule.scope.to_string();
            let base_name = scope.split('(').next().unwrap_or(&scope).to_string();
            let base_name_vo = LayerNameVO::new(&base_name);

            if !result.contains_key(&base_name_vo) {
                continue;
            }

            let scope_vo = LayerNameVO::new(&scope);
            if !result.contains_key(&scope_vo) && scope != base_name {
                if let Some(base_def) = result.get(&base_name_vo) {
                    result.insert(scope_vo.clone(), base_def.clone());
                }
            }

            if scope == LAYER_GLOBAL {
                self.apply_global_rule(&mut result, rule);
            } else if scope.contains('(') {
                self.apply_specialized_rule(&mut result, &scope_vo, rule);
            } else {
                self.apply_base_rule(&mut result, &scope_vo, rule);
            }
        }
        result
    }

    fn apply_global_rule(&self, result: &mut HashMap<LayerNameVO, LayerDefinition>, rule: &ArchitectureRule) {
        let keys: Vec<LayerNameVO> = result.keys().cloned().collect();
        for layer_name in keys {
            if layer_name.to_string().contains('(') {
                continue;
            }
            if let Some(def) = result.get(&layer_name).cloned() {
                result.insert(layer_name, self.merge_rule_into_definition(def, rule));
            }
        }
    }

    fn apply_specialized_rule(&self, result: &mut HashMap<LayerNameVO, LayerDefinition>, scope: &LayerNameVO, rule: &ArchitectureRule) {
        if let Some(def) = result.get(scope).cloned() {
            result.insert(scope.clone(), self.merge_rule_into_definition(def, rule));
        }
    }

    fn apply_base_rule(&self, result: &mut HashMap<LayerNameVO, LayerDefinition>, scope: &LayerNameVO, rule: &ArchitectureRule) {
        if let Some(def) = result.get(scope).cloned() {
            result.insert(scope.clone(), self.merge_rule_into_definition(def, rule));
        }
    }

    fn merge_rule_into_definition(&self, mut def: LayerDefinition, rule: &ArchitectureRule) -> LayerDefinition {
        // Override simple scalar fields
        if rule.word_count.is_some() { def.word_count = rule.word_count; }
        if rule.suffix_policy.is_some() { def.suffix_policy = rule.suffix_policy.clone().unwrap(); }
        if rule.mandatory_import_violation_message.is_some() { def.mandatory_import_violation_message = rule.mandatory_import_violation_message.clone(); }
        if rule.forbidden_import_violation_message.is_some() { def.forbidden_import_violation_message = rule.forbidden_import_violation_message.clone(); }
        if rule.word_count_violation_message.is_some() { def.word_count_violation_message = rule.word_count_violation_message.clone(); }
        if rule.suffix_violation_message.is_some() { def.suffix_violation_message = rule.suffix_violation_message.clone(); }
        if rule.no_primitives_violation_message.is_some() { def.no_primitives_violation_message = rule.no_primitives_violation_message.clone(); }
        if rule.min_lines_violation_message.is_some() { def.min_lines_violation_message = rule.min_lines_violation_message.clone(); }
        if rule.max_lines_violation_message.is_some() { def.max_lines_violation_message = rule.max_lines_violation_message.clone(); }
        if rule.barrel_completeness_violation_message.is_some() { def.barrel_completeness_violation_message = rule.barrel_completeness_violation_message.clone(); }
        if rule.forbid_internal_all_violation_message.is_some() { def.forbid_internal_all_violation_message = rule.forbid_internal_all_violation_message.clone(); }
        if rule.forbidden_bypass_violation_message.is_some() { def.forbidden_bypass_violation_message = rule.forbidden_bypass_violation_message.clone(); }
        if rule.mandatory_class_definition_violation_message.is_some() { def.mandatory_class_definition_violation_message = rule.mandatory_class_definition_violation_message.clone(); }
        if rule.dead_inheritance_bypass_violation_message.is_some() { def.dead_inheritance_bypass_violation_message = rule.dead_inheritance_bypass_violation_message.clone(); }
        if rule.orphan_violation_message.is_some() { def.orphan_violation_message = rule.orphan_violation_message.clone(); }
        if rule.check_unused_mandatory_imports_violation_message.is_some() { def.check_unused_mandatory_imports_violation_message = rule.check_unused_mandatory_imports_violation_message.clone(); }
        if rule.no_domain_logic_violation_message.is_some() { def.no_domain_logic_violation_message = rule.no_domain_logic_violation_message.clone(); }
        if rule.must_implement_service_container_aggregate_violation_message.is_some() { def.must_implement_service_container_aggregate_violation_message = rule.must_implement_service_container_aggregate_violation_message.clone(); }
        if rule.lazy_eager_initialization_only_violation_message.is_some() { def.lazy_eager_initialization_only_violation_message = rule.lazy_eager_initialization_only_violation_message.clone(); }
        if rule.stateless_execution_violation_message.is_some() { def.stateless_execution_violation_message = rule.stateless_execution_violation_message.clone(); }
        if rule.single_execution_goal_violation_message.is_some() { def.single_execution_goal_violation_message = rule.single_execution_goal_violation_message.clone(); }
        if rule.high_level_policy_only_violation_message.is_some() { def.high_level_policy_only_violation_message = rule.high_level_policy_only_violation_message.clone(); }
        if rule.coordinates_multiple_orchestrators_violation_message.is_some() { def.coordinates_multiple_orchestrators_violation_message = rule.coordinates_multiple_orchestrators_violation_message.clone(); }
        if rule.crud_only_violation_message.is_some() { def.crud_only_violation_message = rule.crud_only_violation_message.clone(); }
        if rule.no_decision_logic_violation_message.is_some() { def.no_decision_logic_violation_message = rule.no_decision_logic_violation_message.clone(); }
        if rule.thread_async_safe_violation_message.is_some() { def.thread_async_safe_violation_message = rule.thread_async_safe_violation_message.clone(); }
        if rule.no_domain_data_storage_violation_message.is_some() { def.no_domain_data_storage_violation_message = rule.no_domain_data_storage_violation_message.clone(); }
        if rule.owns_system_health_transitions_violation_message.is_some() { def.owns_system_health_transitions_violation_message = rule.owns_system_health_transitions_violation_message.clone(); }
        if rule.lifecycle_tracking_only_violation_message.is_some() { def.lifecycle_tracking_only_violation_message = rule.lifecycle_tracking_only_violation_message.clone(); }
        if rule.forbid_any_type_violation_message.is_some() { def.forbid_any_type_violation_message = rule.forbid_any_type_violation_message.clone(); }

        // Override boolean fields
        if rule.no_primitives != BooleanVO::new(false) { def.no_primitives_no_primitives_hack(rule.no_primitives); }
        if !rule.barrel_completeness.value { def.barrel_completeness = rule.barrel_completeness; }

        // Simpler approach: just override basic fields that are simple
        // Collection fields
        if rule.allowed_import.is_some() { def.allowed_import = rule.allowed_import.clone().unwrap(); }
        if rule.forbidden_import.is_some() { def.forbidden_import = rule.forbidden_import.clone().unwrap(); }
        if rule.mandatory_import.is_some() { def.mandatory_import = rule.mandatory_import.clone().unwrap(); }
        if rule.allowed_suffix.is_some() { def.allowed_suffix = rule.allowed_suffix.clone().unwrap(); }
        if rule.forbidden_suffix.is_some() { def.forbidden_suffix = rule.forbidden_suffix.clone().unwrap(); }
        if rule.exceptions.is_some() { def.exceptions = rule.exceptions.clone().unwrap(); }
        if rule.forbidden_bypass.is_some() { def.forbidden_bypass = rule.forbidden_bypass.clone().unwrap(); }
        if rule.orphan_entry_points.is_some() { def.orphan_entry_points = rule.orphan_entry_points.clone().unwrap(); }

        def
    }
}
